### PR TITLE
Buffer mongo log record insertions

### PR DESCRIPTION
**Pull request by axw**
## Description of change

- Refactor and simplify the state.DbLogger code.
- Introduce a state/logdb package, which I intend to move the existing DbLogger and friends into, but first things first. This package initially holds a BufferedLogger, which buffers log records and flushes to an underlying logger (e.g. DbLogger) when either the buffer capacity or a time interval are exceeded.
- Update state.DbLogger.Log to insert records in bulk.
- Update the "release" function returned by StatePool.Get, and StatePool.Remove, to report whether or not they remove/close the State.
- Update the apiserver logsink/logtransfer code to share a single DbLogger for all uses of a shared state pool entry, to cut down on the number of sockets used for logging.
- Update the apiserver logsink/logtransfer code to use state/logdb.BufferedLogger to buffer log writes. The parameters are currently hard-coded: a buffer capacity of 1024, and flush interval of 2 seconds. We can make these configurable as needed.

Note to reviewers: there are a few refactoring commits, which you may wish to review first. If it makes your life easier, I'll split this up - I just wanted to get it in front of people ASAP.

## QA steps

1. juju bootstrap localhost
2. juju add-machine
3. juju debug-log -m controller

Watch the logs roll in; perform some activity and observe that, unless there is a flurry of logs, debug-log will take ~2 seconds to update. The number of mongo sockets and insertions should be reduced from before. I added temporary instrumentation to prove this to myself, but it should be possible to eyeball the improvements via mongo stats with a larger deployment.

## Documentation changes

None.

## Bug reference

None.

### CODE REVIEW

**Comment by jameinel** [comment]
Do we have concrete information that this is better?
If we changed logging to aggregate into time-based log objects (rather than
every log being a separate object), then this would seem like a bigger win.
And while Insert(100)*10 is probably better than Insert(1)*1000, it would
be good to have a concrete understanding of what the buffering is going to
get us.

File: apiserver/debuglog.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ type debugLogHandler struct {
```
Comment (jameinel) [inline apiserver/debuglog.go]: no actual changes here, just a *struct to struct, right?

File: apiserver/apiserver.go [hunk 2]
```diff
@@ -596,7 +599,7 @@ func (srv *Server) endpoints() []apihttp.Endpoint {
 	)
 
 	add("/model/:modeluuid/applications/:application/resources/:resource", &ResourcesHandler{
-		StateAuthFunc: func(req *http.Request, tagKinds ...string) (ResourcesBackend, func(), names.Tag, error) {
+		StateAuthFunc: func(req *http.Request, tagKinds ...string) (ResourcesBackend, state.StatePoolReleaser, names.Tag, error) {
```
Comment (jameinel) [inline apiserver/apiserver.go]: This is also just a drive-by, right? not related to dblogger changes? Just saw we weren't using the state pool and releasing them?

File: apiserver/debuglog_db.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ func newDebugLogDBHandler(ctxt httpContext) http.Handler {
 
 func handleDebugLogDBRequest(
 	st state.LogTailerState,
-	reqParams *debugLogParams,
+	reqParams debugLogParams,
```
Comment (jameinel) [inline apiserver/debuglog_db.go]: I assume the same thing here, just changing from *struct to struct, along with the tests.

If otherwise, please ask for a deeper review.

File: apiserver/httpcontext.go [hunk 4]
```diff
@@ -116,7 +120,10 @@ func checkPermissions(tag names.Tag, acceptFunc common.GetAuthFunc) (bool, error
 // has admin permissions on the controller model. The method also gets the
 // model uuid for the model being migrated from a request header, and returns
 // the state instance for that model.
-func (ctxt *httpContext) stateForMigration(r *http.Request, requiredMode state.MigrationMode) (st *state.State, returnReleaser func(), err error) {
+func (ctxt *httpContext) stateForMigration(
+	r *http.Request,
```
Comment (jameinel) [inline apiserver/httpcontext.go]: nit: seems odd to wrap the params but not the returns

File: apiserver/logsink.go [hunk 5]
```diff
@@ -7,27 +7,105 @@ import (
...
 	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/logdb"
+)
+
+const (
+	// dbLoggerBufferSize is the capacity of the log buffer.
+	// When the buffer fills up, it will be flushed to the
+	// database.
+	dbLoggerBufferSize = 1024
```
Comment (jameinel) [inline apiserver/logsink.go]: how did we derive this constant? was there testing that 10k was too big, 100 was too small?
Should we try to expose knobs in agent.conf as we did for other arbitrary constants that we may want to tune?

File: apiserver/logsink.go [hunk 6]
```diff
@@ -7,27 +7,105 @@ import (
...
+	defer d.mu.Unlock()
+	if l, ok := d.loggers[st]; ok {
+		l.Close()
+		delete(d.loggers, st)
+	}
+}
+
+// dispose closes all dbloggers in the map, and clears the memory. This
+// must not be called concurrently with any other dbloggers methods.
+func (d *dbloggers) dispose() {
```
Comment (jameinel) [inline apiserver/logsink.go]: disposeAll ?

File: apiserver/logsink.go [hunk 7]
```diff
@@ -61,24 +143,44 @@ func (s *agentLoggingStrategy) init(ctxt httpContext, req *http.Request) error {
...
 		return errors.Trace(err)
 	}
-	s.releaser = releaser
 	s.version = ver
 	s.entity = entity.Tag()
 	s.filePrefix = st.ModelUUID() + ":"
-	s.dbLogger = state.NewEntityDbLogger(st, s.entity, s.version)
+	s.dblogger = s.dbloggers.get(st)
+	s.releaser = func() {
+		if removed := releaseState(); removed {
```
Comment (jameinel) [inline apiserver/logsink.go]: so is the return value of releaseState "has this model been deleted" ?

File: state/logdb/buf.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	defer b.mu.Unlock()
+	for len(in) > 0 {
+		r := cap(b.buf) - len(b.buf)
+		n := len(in)
+		if n > r {
+			n = r
+		}
+		b.buf = append(b.buf, in[:n]...)
+		in = in[n:]
+		if len(b.buf) == cap(b.buf) {
```
Comment (jameinel) [inline state/logdb/buf.go]: while this is arguably correct, I'm always a fan of ">=" just to avoid "something caused it to go past the target, buffer everything forever after and never stop"

File: state/logdb/buf.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	}
+}
+
+// Log is part of the Logger interface.
+//
+// BufferedLogger's Log implementation will buffer log records up to
+// the specified capacity and duration; after either of which is exceeded,
+// the records will be flushed to the underlying logger.
+func (b *BufferedLogger) Log(in []state.LogRecord) error {
+	b.mu.Lock()
```
Comment (jameinel) [inline state/logdb/buf.go]: I'm curious if the logic around this would be clearer (or more confusing) if Log just sent a message onto a channel for a worker loop to consume, rather than having locks, etc.
Most of Juju code is based around worker loops.

File: state/logdb/buf.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+func (b *BufferedLogger) flushOnTimer() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	// Can't do anything about errors here, except to
+	// ignore them and let the Log() method report them
+	// when the buffer fills up.
+	b.flush()
+}
+
+func (b *BufferedLogger) flush() error {
```
Comment (jameinel) [inline state/logdb/buf.go]: a comment here about "flush assumes you are already holding the mutex" would be helpful for readers

File: state/logdb/buf.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+		}
+		b.buf = append(b.buf, in[:n]...)
+		in = in[n:]
+		if len(b.buf) == cap(b.buf) {
+			if err := b.flush(); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	if len(b.buf) > 0 && b.flushTimer == nil {
```
Comment (jameinel) [inline state/logdb/buf.go]: This generally feels like a re-usable pattern that would be nice to use for Pinger. It is independent of *what* buf is and what function is being called wrt flushing the buffer.
TimeFlushingBuffer ?
Its not a lot of code, so maybe its easier to have it inlined, but it does seem like a pattern we're going to be repeating.

File: state/logdb/buf_test.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	select {
+	case records := <-s.mock.called:
+		c.Assert(records, jc.DeepEquals, in)
+		s.mock.CheckCalls(c, []testing.StubCall{
+			{"Log", []interface{}{in}},
+		})
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out waiting for logs to be flushed")
+	}
+}
```
Comment (jameinel) [inline state/logdb/buf_test.go]: The other parts you are missing are that it *doesn't* trigger any future logs
and what time we will be flushing after future logging. (the set back to nil and it getting set again once we see another log)

Also, we are missing a test that if you Log([]1010) items, then we will flush the 1000 immediately, but wait the 1minute to flush the remaining 10, but we guarantee the 10 will get flushed eventually.

And potentially the "log 2 items, then N > buffer". Maybe that is just combined with the previous test. But the general "some stuff is queued, we overflow the buffer, but still have stuff queued".

File: state/logs.go [hunk 13]
```diff
@@ -216,20 +216,50 @@ func NewDbLogger(st ModelSessioner) *DbLogger {
...
+// DbLogger is scoped to a single model, and ID is
+// controlled by the DbLogger code.
+func (logger *DbLogger) Log(records []LogRecord) error {
+	for _, r := range records {
+		if err := validateInputLogRecord(r); err != nil {
+			return errors.Annotate(err, "validating input log record")
+		}
+	}
+	// TODO(axw) copy session here and close after?
+	bulk := logger.logsColl.Bulk()
```
Comment (jameinel) [inline state/logs.go]: If we have a separate logger per State object, I think re-using just the same session is better than scaling sessions per Log calls.

File: state/logs.go [hunk 14]
```diff
@@ -216,20 +216,50 @@ func NewDbLogger(st ModelSessioner) *DbLogger {
...
+	for _, r := range records {
+		var versionString string
+		if r.Version != version.Zero {
+			versionString = r.Version.String()
+		}
+		bulk.Insert(&logDoc{
+			// TODO(axw) Use a controller-global int
+			// sequence for Id, so we can order by
+			// insertion.
+			Id:       bson.NewObjectId(),
```
Comment (jameinel) [inline state/logs.go]: object ids have a counter-per state and an embedded timestamp, not sufficient? I guess the timestamp only has 1s resolution which is why we have nano timestamps. (probably ms is enough, but something more than 1s)

File: state/logs.go [hunk 15]
```diff
@@ -445,7 +440,7 @@ func (t *logTailer) processReversed(query *mgo.Query) error {
 		return errors.Errorf("too many lines requested (%d) maximum is %d",
 			t.params.InitialLines, maxInitialLines)
 	}
-	query.Sort("-e", "-t", "-_id")
+	query.Sort("-t", "-_id")
```
Comment (jameinel) [inline state/logs.go]: the index we have required e t i, did that change with the split out but we didn't change the query?

File: state/pool.go [hunk 16]
```diff
@@ -104,33 +111,34 @@ func (p *StatePool) Get(modelUUID string) (*State, func(), error) {
 // release indicates that the client has finished using the State. If the
 // state has been marked for removal, it will be closed and removed
 // when the final Release is done.
```
Comment (jameinel) [inline state/pool.go]: update this comment for what the boolean means?

File: apiserver/debuglog.go (hunk 1 above)
Comment (axw) [inline apiserver/debuglog.go]: Yep, there's a few refactoring commits up front.

File: apiserver/apiserver.go (hunk 2 above)
Comment (axw) [inline apiserver/apiserver.go]: This is from https://github.com/juju/juju/pull/7502/commits/5bc3e7a7981ffed96fc130986ee24e7b34a12943, but no functional change in this part of the code. The StatePoolReleaser has a boolean result that tells the caller whether or not releasing caused the State object to be closed and removed from the pool.

File: apiserver/debuglog_db.go (hunk 3 above)
Comment (axw) [inline apiserver/debuglog_db.go]: Yep.

File: apiserver/logsink.go (hunk 5 above)
Comment (axw) [inline apiserver/logsink.go]: It's ~the same as what's used in the juju/txn pruning code. I did do some testing, and saw that in a small deployment the buffer was basically never reached, which means we'll generally wait 2s to flush. Yes, I think we'll want to expose a tuning knob, but I'll probably do that in a followup. I'll do some more analysis first.

File: apiserver/logsink.go (hunk 6 above)
Comment (axw) [inline apiserver/logsink.go]: "dispose of the dbloggers" is how it reads to me

File: apiserver/logsink.go (hunk 7 above)
Comment (axw) [inline apiserver/logsink.go]: more that the *State object is no longer usable, but effectively yes; because we don't remove them until the model is dead

File: state/logdb/buf.go (hunk 9 above)
Comment (axw) [inline state/logdb/buf.go]: I did consider doing it that way, but having a goroutine looping for work means an always-on timer or a more complicated condition variable or something. Using a timer means waking up and burning CPU to find there's nothing to do. Probably doesn't matter much, but I'd like to get away from doing it where we can, and this is pretty simple and contained.

File: state/logdb/buf.go (hunk 8 above)
Comment (axw) [inline state/logdb/buf.go]: Sure, done.

File: state/logdb/buf.go (hunk 11 above)
Comment (axw) [inline state/logdb/buf.go]: It might be possible to do with an interface, but I can't see how at the moment. IMO it's simple and small enough to inline.

File: state/logdb/buf.go (hunk 10 above)
Comment (axw) [inline state/logdb/buf.go]: Done.

File: state/logdb/buf_test.go (hunk 12 above)
Comment (axw) [inline state/logdb/buf_test.go]: Done.

File: state/logs.go (hunk 13 above)
Comment (axw) [inline state/logs.go]: Agreed, I should have dropped that - will do so now.

File: state/logs.go (hunk 14 above)
Comment (axw) [inline state/logs.go]: It needs to be monotonic in the face of HA, so not sufficient if we want to preserve order exactly. There may be an alternative to this, I was just updating the name of the person responsible.

File: state/logs.go (hunk 15 above)
Comment (axw) [inline state/logs.go]: Yeah, we no longer store the model UUID (what was "e") as part of the record. Each model's logs now go into a separate collection.

File: state/pool.go (hunk 16 above)
Comment (axw) [inline state/pool.go]: Done.

**Comment by axw** [comment]
> Do we have concrete information that this is better?
If we changed logging to aggregate into time-based log objects (rather than
every log being a separate object), then this would seem like a bigger win.
And while Insert(100)*10 is probably better than Insert(1)*1000, it would
be good to have a concrete understanding of what the buffering is going to
get us.

This PR is just about improving the performance of log insertion, and I don't see how aggregating into time-based objects would help that. That would be helpful for querying and pruning.

I've eyeballed a definite CPU usage reduction, but I'll run again with Prometheus and Telegraf and provide a graph.

**Comment by axw** [comment]
The graphs aren't terribly easy to read without the annotations and such, so here's some highlights. I deployed 10 units of "fill-logs" from the acceptance tests charm repo, and ran "fill-machine" with 500MB on each machine, 10 times in a row, with controller model's log level set to TRACE. This is not supposed to be super realistic, just driving logging up high enough that it impacts CPU.

 - without my changes, number of in-use sockets sits at about 30; with my changes, it fluctuates between 6 and 7.
 - without my changes, CPU ranges between 25-30% user, 6-7% system; with my changes ranges between 20-25% user, 3-5% system. Roughly half iowait% with my changes.
 - without my changes, system load 1 min average is ~1.8; with my changes, just under 1

**Comment by axw** [comment]
!!build!!

File: worker/dblogpruner/worker_test.go [hunk 17]
```diff
@@ -140,11 +140,19 @@ func (s *suite) TestPrunesLogsBySize(c *gc.C) {
 }
 
 func (s *suite) addLogs(c *gc.C, t0 time.Time, text string, count int) {
-	dbLogger := state.NewEntityDbLogger(s.state, names.NewMachineTag("0"), version.Current)
+	dbLogger := state.NewDbLogger(s.state) // names.NewMachineTag("0"), version.Current)
```
Comment (wallyworld) [inline worker/dblogpruner/worker_test.go]: delete me

Comment (axw) [inline worker/dblogpruner/worker_test.go]: fixed

**Comment by axw** [comment]
$$merge$$

File: apiserver/logsink.go (hunk 5 above)
Comment (jameinel) [inline apiserver/logsink.go]: I should note that Bulk has a limit of 1000 records (according to txn pruning) so we probably want to limit the chunk size slightly smaller.

File: state/logdb/buf.go (hunk 9 above)
Comment (jameinel) [inline state/logdb/buf.go]: You can actually do the same tricks of not having the timer active when there isn't work, etc.
That said, in anything where we're actually active and care about the buffering, do we really think we'll be burning much CPU on timers checking?

File: apiserver/logsink.go [hunk 18]
```diff
@@ -7,27 +7,105 @@ import (
...
+const (
+	// dbLoggerBufferSize is the capacity of the log buffer.
+	// When the buffer fills up, it will be flushed to the
+	// database.
+	dbLoggerBufferSize = 1024
+
+	// dbLoggerFlushInterval is the amount of time to allow
+	// a log record to sit in the buffer before being flushed
+	// to the database.
+	dbLoggerFlushInterval = 2 * time.Second
```
Comment (axw) [inline apiserver/logsink.go]: Both are configurable in https://github.com/juju/juju/pull/7505.
